enc.go

```diff
@@ -89,22 +90,37 @@ func EncoderForType(encType string) Encoder {
 // Publish publishes the data argument to the given subject. The data argument
 // will be encoded using the associated encoder.
 func (c *EncodedConn) Publish(subject string, v interface{}) error {
+	ctx := context.Background()
```

Comment:
I just ran a bench between your branch and master and on pure publish (nats-bench -np 1 -ns 0 -n 10000000 foo) I see a 10% drop. The issue I have is that you get a context here and call PublishWithContext(), but if there is no publish wrapper registered, we end-up calling `publish(*Conn, context.Context, ...)` but we don't do anything with the context.
So at the very least I would refactor all that and delay getting context.Background() in case where a context is not provided and we *know* that we are going to invoke some `PublishFunc` functions. Does that make sense?

Reply:
@kozlovic Great idea! That completely makes sense. Thanks for running those benchmarks.

Reply:
I also wonder the usefulness of the context.Context in this context (pun intended). In the examples you provide, you don't really make use of it. And I am not sure the publish calls can be cancelled (assuming you provide your own context, not context.Background() which can't be cancelled).

Reply:
@kozlovic pun appreciated!

Thanks for the question, let me be more clear and giving an example of how we'd be instructing our users to instrument their Publish calls.

Here's what I'm thinking we'd give our users for a `PublishWrapper`. It will look in the context for our `newrelic.Transaction` and use it to start a new segment for the Publish call.

```go
func NRWrapper(fn nats.PublishFunc) nats.PublishFunc {
    return func(nc *nats.Conn, ctx context.Context, subj, reply string, data []byte) error {
        if txn := newrelic.FromContext(ctx); nil != txn {
            seg := newrelic.ExternalSegment{
                StartTime: newrelic.StartSegmentNow(txn),
                URL:       nc.ConnectedUrl(),
                Procedure: "Publish/" + subj,
                Library:   "NATS",
            }   
            defer seg.End()
        }   
        return fn(nc, ctx, subj, reply, data)
    }   
}
```

Then, as long as they have registered this wrapper and given a context with the transaction in it to the publish call, they will get all the magic of our New Relic metrics and segments. Without the context, we'd be unable to know which transaction the segment should belong to.

Hope this helps 😃